macaroonbakery/bakery.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2017 Canonical Ltd.
+# Licensed under the LGPLv3, see LICENCE file for details.
+
+import base64
+import json
+import webbrowser
+
+import macaroons
+import requests
+
+
+ERR_INTERACTION_REQUIRED = 'interaction required'
+ERR_DISCHARGE_REQUIRED = 'macaroon discharge required'
+TIME_OUT = 30
+
+
+class BakeryAuth():
+    def __init__(self, visit_page=None, key=None, bakery_protocol='1'):
+        if visit_page is None:
+            visit_page = _visit_page_with_browser
+        self._visit_page = visit_page
+        self._key = key
+        self._protocol_version = bakery_protocol
+
+    def __call__(self, req):
+        hook = _prepare_discharge_hook(req.copy())
+        req.headers['Bakery-Protocol-Version'] = self._protocol_version
+        req.register_hook(event='response', hook = hook)
+        return req
+
+
+def _prepare_discharge_hook(req):
+    def hook(response, *args, **kwargs):
+        if response.status_code != 401 and response.status_code != 407:
+            return response
+        if response.headers['Content-Type'] != 'application/json':
+            return response
+        error = response.json()
+        if error['Code'] == ERR_DISCHARGE_REQUIRED:
+            macaroon = _deserialize(error['Info']['Macaroon'])
+            discharges = discharge_all_with_key(macaroon)
+            encoded_discharges = map(_serialize_macaroon_string, discharges)
+            macaroons = base64.urlsafe_b64encode(
+                '[' + ','.join(encoded_discharges) + ']')
+            s = requests.Session()
+            req.headers['Bakery-Protocol-Version'] = '1'
```

Comment:
I'd set this version in the BakeryAuth before passing the copy of the request, so that we can use the provided protocol version.